### PR TITLE
Add settings `output_format_orc_dictionary_key_size_threshold` to allow user to enable dict encoding for string column in ORC output format

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add settings `output_format_orc_dictionary_key_size_threshold` to allow user to enable dict encoding for string column in ORC output format. It helps reduce the output orc file size and improve reading performance significantly.  

TODO
- [x] Support `EncodedStringVectorBatch` for interface `StringColumnWriter::add`  (in https://github.com/ClickHouse/orc/pull/15)

1. Write ORC file with string dict encoding enabled
``` sql
set output_format_orc_dictionary_key_size_threshold = 1;
select concat('gluten ', cast(rand()%1000 as String)) from numbers(10000000) into outfile 'dict.orc' truncate;
10000000 rows in set. Elapsed: 2.794 sec. Processed 10.00 million rows, 80.00 MB (3.58 million rows/s., 28.63 MB/s.)
Peak memory usage: 9.07 MiB.
```

2. Write ORC file with string dict encoding disabled 
``` sql
set output_format_orc_dictionary_key_size_threshold = 0; 
select concat('gluten ', cast(rand()%1000 as String)) from numbers(10000000) into outfile 'no-dict.orc' truncate;
10000000 rows in set. Elapsed: 1.222 sec. Processed 10.00 million rows, 80.00 MB (8.18 million rows/s., 65.46 MB/s.)
Peak memory usage: 9.07 MiB.
```

3. Compare ORC file size
``` bash
$ ll -rt *dict*
-rw-r--r-- 1 root root 22731823 Aug 20 15:42 no-dict.orc
-rw-r--r-- 1 root root 14921714 Aug 20 15:42 dict.orc
```

4. Compare reading performance of both ORC files
``` sql
select * from file('dict.orc') format Null;
0 rows in set. Elapsed: 0.166 sec. Processed 10.00 million rows, 14.92 MB (60.19 million rows/s., 89.82 MB/s.)

select * from file('no-dict.orc') format Null;
0 rows in set. Elapsed: 0.385 sec. Processed 10.00 million rows, 22.73 MB (25.96 million rows/s., 59.00 MB/s.)
```

5. Conclusions
- Writing string column with dict encoding is 2.3x slower than without dict encoding. But I improved its performance in https://github.com/ClickHouse/orc/pull/15. Hope for your review. 
- ORC files written with dict encoding is 1.5x smaller than without dict encoding
- Reading string column with dict encoding is 2.3x faster than without dict encoding 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [x]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 03ab6252653ddcb70d1fef0a340bfb52a8d63f89 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68591/03ab6252653ddcb70d1fef0a340bfb52a8d63f89/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Core/SettingsChangesHistory.cpp [hunk 1]
```diff
@@ -94,7 +94,8 @@ static std::initializer_list<std::pair<ClickHouseVersion, SettingsChangesHistory
             {"type_json_skip_duplicated_paths", false, false, "Allow to skip duplicated paths during JSON parsing"},
             {"join_output_by_rowlist_perkey_rows_threshold", 0, 5, "The lower limit of per-key average rows in the right table to determine whether to output by row list in hash join."},
             {"allow_experimental_vector_similarity_index", false, false, "Added new setting to allow experimental vector similarity indexes"},
-            {"input_format_try_infer_datetimes_only_datetime64", true, false, "Allow to infer DateTime instead of DateTime64 in data formats"}
+            {"input_format_try_infer_datetimes_only_datetime64", true, false, "Allow to infer DateTime instead of DateTime64 in data formats"},
+            {"output_format_orc_dictionary_key_size_threshold", 0.0, 0.0, "For a string column in ORC output format, if the number of distinct values is greater than this fraction of the total number of non-null rows, turn off dictionary encoding. Otherwise dictionary encoding is enabled"},
```
Comment (Avogar) [inline src/Core/SettingsChangesHistory.cpp:98]: Please, move it to 24.9 section

**Review by Avogar — APPROVED** [review]

**Comment by taiyang-li** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/68591/0de3b1dacbc587e7f6789dec3fe92e276f332f56/stateful_tests__ubsan_.html  I can't figure out why it failed

**Comment by devcrafter** [comment]
> https://s3.amazonaws.com/clickhouse-test-reports/68591/0de3b1dacbc587e7f6789dec3fe92e276f332f56/stateful_tests__ubsan_.html I can't figure out why it failed

Please update the PR with recent master. It was an issue in CI, - was fixed by https://github.com/ClickHouse/ClickHouse/pull/69483/

**Pull request merged**